extensions/spring-cloud-config-client/runtime/src/main/java/io/quarkus/spring/cloud/config/client/runtime/credentials/SpringCloudConfigClientOidcCredentialsProvider.java

```diff
@@ -0,0 +1,110 @@
+package io.quarkus.spring.cloud.config.client.runtime.credentials;
+
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import io.quarkus.oidc.client.OidcClient;
+import io.quarkus.oidc.client.OidcClientConfig;
+import io.quarkus.oidc.client.runtime.OidcClientImpl;
+import io.quarkus.oidc.common.runtime.OidcCommonConfig;
+import io.quarkus.oidc.common.runtime.OidcCommonUtils;
+import io.quarkus.oidc.common.runtime.OidcConstants;
+import io.quarkus.spring.cloud.config.client.runtime.SpringCloudClientCredentialsProvider;
+import io.quarkus.spring.cloud.config.client.runtime.SpringCloudConfigClientConfig;
+import io.vertx.mutiny.core.MultiMap;
+import io.vertx.mutiny.ext.web.client.HttpRequest;
+import io.vertx.mutiny.ext.web.client.WebClient;
+
+public class SpringCloudConfigClientOidcCredentialsProvider implements SpringCloudClientCredentialsProvider {
+
+    private static final String OIDC_CLIENT_ID = "SpringCloudConfigOidcClient";
+
+    OidcClient oidcClient;
+
+    public SpringCloudConfigClientOidcCredentialsProvider(SpringCloudConfigClientConfig clientConfig, WebClient webClient) {
+        OidcClientConfig oidcClientConfig = clientConfig.oidc;
+
+        if (!oidcClientConfig.getId().isPresent()) {
+            oidcClientConfig.setId(OIDC_CLIENT_ID);
+        }
+
+        String authServerUriString = OidcCommonUtils.getAuthServerUrl(oidcClientConfig);
+        String tokenRequestUri;
+        if (!oidcClientConfig.discoveryEnabled) {
+            tokenRequestUri = OidcCommonUtils.getOidcEndpointUrl(authServerUriString, oidcClientConfig.tokenPath);
+        } else {
+            tokenRequestUri = OidcCommonUtils
+                    .discoverMetadata(webClient, authServerUriString, clientConfig.connectionTimeout.toMillis())
+                    .onItem().transform(json -> json.getString("token_endpoint")).await().indefinitely();
+        }
+
+        String grantType = oidcClientConfig.grant.getType().getGrantType();
+        MultiMap grantClientParams = getGrantClientParams(oidcClientConfig, grantType);
+
+        MultiMap tokenGrantParams = getTokenGrantParam(oidcClientConfig, grantType);
+
+        this.oidcClient = new OidcClientImpl(webClient, tokenRequestUri, grantType, tokenGrantParams, grantClientParams,
+                oidcClientConfig);
+    }
+
+    @Override
+    public void addAuthenticationInfo(HttpRequest<?> request) {
+        request.bearerTokenAuthentication(oidcClient.getTokens().await().indefinitely().getAccessToken());
```

Comment:
I'm not 100% sure requesting a new token on every authentication request will work with the Spring Cloud provider - as there are usually some limits on how many tokens a given client can acquire.

The produced OidcClient in `quarkus-oidc-client` will cache the token and handle auto-refreshing it once they have expired - but if the client is used directly then the token has to be refreshed if it has expired.

I think you should use `io.quarkus.oidc.client.runtime.TokensHelper` instead (this is used by OidcClient producers) - pass it the initialized OidcClient - and it will take care of the rest (caching, refreshing, etc)

Reply:
You are right, I used the `TokensHelper` here. However, I think this is called only once when the application starts, it load config then, it will only be refreshed on the next restart.

Reply:
@glefloch Hi, I meant the access token will be refreshed if needed - it will be acquired and then reused while it is active and then refreshed when needed...